ninjabackend: Try to guess library dependencies for linker invocation.

textshell opened the pull request:
The linkers currently do not support ninja compatible output of
dependencies used while linking. Try to guess which files will be used
while linking in python code and generate conservative dependencies to
ensure changes in linked libraries are detected.
    
This generates dependencies on the best match for static and shared
linking, but this should not be a problem, except for spurious
rebuilding when only one of them changes, which should not be a problem.
    
Also makes sure to ignore any libraries generated inside the build, to
keep the optimisation working where changes in a shared library only
cause relink if the symbols have changed as well.

jpakkane commented:
Looks reasonable. @nirbheek, what you think?

textshell commented:
@nirbheek Any news on this?

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2387,6 +2387,73 @@ def get_link_whole_args(self, linker, target):
+        it = iter(commands)
+        try:
+            while True:
+                item = next(it)
```
Why are you converting the list to an iterator and then looping over it anyway?

textshell replied:
Sometimes the code needs 2 items. i.e. '-L', '/some/path'. I think having an iterator to pull from is easier than doing the more like a state machine.

nirbheek replied:
... actually maybe we should warn when there's no argument after an empty `-L`, so we don't want to silently `pass` in that case. Or maybe an error, because this will cause weird bugs where the next argument gets incorrectly slurped as a library search directory.

textshell replied:
I wrote this code with the assumption that it should never break a build that worked before, because this is a bit of a second guessing game with the linker. The linker is the actual point of truth.

Not sure what mesons policy on warning is. But i could add a working if the commands end with an unmatches `-L` or `-l`

textshell replied:
I've added a warning and refactored the looping code. I didn't know that a iterator is valid for a `for` loop as well as an iterable, but it seems to work.

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2387,6 +2387,73 @@ def get_link_whole_args(self, linker, target):
+
+        guessed_dependencies = []
+
+        if hasattr(linker, 'get_library_naming'):
```
Which linkers don't implement this? Should we perhaps implement this for those instead?

textshell replied:
`linker` is actually a compiler iirc. And i think only the C family compilers (c, cpp plus their objective variants) have this.

If we want to change that i would prefer to discuss that in a separate ticket.

nirbheek replied:
Also please add a TODO comment about this, with a list of compilers that are affected.

textshell replied:
I don't think there is much work that short-circuiting could avoid. We need to parse the command line to get the absolute paths. And everything else is inside this if.

nirbheek replied:
All right.

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2387,6 +2387,73 @@ def get_link_whole_args(self, linker, target):
+
+    def guess_external_link_dependencies(self, linker, target, commands, internal):
+        # This is a work around for missing dependency file generation in the linker
+        # https://sourceware.org/bugzilla/show_bug.cgi?id=22843
```
Is there a corresponding bug for `lld`? FreeBSD will finish its migration to `lld` this year, and there are other projects already using it. `lld` might be our best hope for implementing a better future.

textshell replied:
I might need to rework this comment. Because even fixing the binutils bug, meson could not use that without regressions. Because for libraries in the same project meson does further optimization by generating symbol files. That of course needs dependencies managed by meson by adding ninja dependencies that use an additional check inbetween.

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2387,6 +2387,73 @@ def get_link_whole_args(self, linker, target):
+                    else:
+                        libs.append(next(it))
+
+                if os.path.isabs(item) and self.environment.is_library(item) and os.path.isfile(item):
```
Same, `continue` or `elif`.

textshell commented:
I think i should have added fixes for all review comments now.
The test uses static libraries only for now, because that is the most important case.
Also testing shared libraries would certainly be better, but maybe that takes to much CI time.

I've rewritten the comment to also explain why the linker outputing a ninja compatible dependencies file does not trivially replace this code.

ignatenkobrain commented on run_unittests.py:
```diff
@@ -1909,6 +1909,48 @@ def test_ndebug_if_release_enabled(self):
+        env = Environment(testdirlib, self.builddir, self.meson_command,
+                          get_fake_options(self.prefix), [])
+        if env.detect_c_compiler(False).get_id() != 'msvc': # default_library=static doesn't work with msvc
+            extra_args='-Ddefault_library=static'
```
__[Flake8]__

[E225] missing whitespace around operator


<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on run_unittests.py:
```diff
@@ -1909,6 +1909,56 @@ def test_ndebug_if_release_enabled(self):
+        for tmp in os.scandir(self.builddir):
+            print(tmp.path, tmp.stat().st_mtime_ns)
+        self.assertRebuiltTarget('app')
+        0/0
```
__[Flake8]__

[E226] missing whitespace around arithmetic operator


<!-- Comment by [SideCI](https://sideci.com) -->

textshell commented:
sorry for spamming this PR with trial commits. But it seems i finally got a unit test setup that also works with the various windows compiler variants. (And added a tiny fix for msvc)

So this PR should now be good from my PoV, please rereview. This currently still contains one appveyor hack that i need to remove before merging for the thankfully now fixed artifact upload failures. My plan if nobody yells for more time is to squash the "fixup!" comments and remove '.appveyor.yml: Disable PushArtifact for now. Appveyor backend seems to be broken at the moment.' sometime friday. I don't want to force push right now, to give travis some time to find an OS X slot, although OS X always worked when linux was ok.

textshell commented:
@nirbheek Any news on this?

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2456,6 +2456,72 @@ def get_link_whole_args(self, linker, target):
+        it = iter(commands)
+        for item in it:
+            if item in internal:
+                continue
```
Won't this skip if `item` is `-L` and that's already in `internal`  as `['-L', 'foo']` and then cause the library path in the next item to be dropped?

textshell replied:
`internal` should never contain any options. It's generated by meson for internal dependencies (i.e. output files of targets). And those always are pathes.
But i'll add an additional condition so this only happens when `not item.startswith('-')` as the compiler driver will not use argument starting with dash as direct inputs anyway.

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2456,6 +2456,72 @@ def get_link_whole_args(self, linker, target):
+                    try:
+                        path = next(it)
+                    except StopIteration:
+                        mlog.warning("Generated linker command has -L argument without following path")
```
Exit loop now since we're done? Otherwise `path` will be undefined and will cause a traceback on the next statement.

textshell replied:
indeed. I'll fix this for both cases.

nirbheek commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2456,6 +2456,72 @@ def get_link_whole_args(self, linker, target):
+                    try:
+                        libs.append(next(it))
+                    except StopIteration:
+                        mlog.warning("Generated linker command has '-l' argument without following library name")
```
Ditto.

nirbheek commented on run_unittests.py:
```diff
@@ -1923,6 +1923,50 @@ def test_ndebug_if_release_enabled(self):
+        extra_args = None
+        env = Environment(testdirlib, self.builddir, self.meson_command,
+                          get_fake_options(self.prefix), [])
+        if env.detect_c_compiler(False).get_id() != 'msvc': # default_library=static doesn't work with msvc
```
Why doesn't it work with MSVC? What went wrong there?

textshell replied:
It does not build a static library at all (#3331). Maybe knowing now that this should be fixable by using `--default-library=static` i could retry that. On the other hand this tests that this also works with dynamic libraries without blowing up the testing times. Maybe i should just remove that comment.

textshell replied:
Ok. I pushed an updated commit. I hope it still passes appveyor, because i can't test windows locally.

textshell commented:
Thanks for reviewing.
I've pushed an amended commit with fixes for all three comments.
If you want to review the diff see https://github.com/mesonbuild/meson/commit/39d8afec31cca0386db7a43a11d7cc6491737bc7

textshell commented:
I had to revert to testing with a shared library on windows again. Building static now worked, but installed static libraries on msvc can't be used with -lname, because how windows library naming with meson works. But i want to test the library resolving code path so i switched back to shared library on windows and updated the comment as follows:

```
# static libraries are not linkable with -l with msvc because meson installs them
# as .a files which unix_args_to_native will not know as it expects libraries to use
# .lib as extension. For a DLL the import library is installed as .lib. Thus for msvc
# this tests needs to use shared libraries to test the path resolving logic in the
# dependency generation code path.
```
